<div style="display: flex; align-items: center; gap: 20px; font-family: Arial, sans-serif;">
    <div style="margin-right: 30px;">
        <img src="../Sources/Logo.png" width="300" style="border-radius: 10px;"/>
    </div>
    <div style="line-height: 1.5;">
        <strong style="font-size: 18px; display: block; margin-bottom: 5px;">
            <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">
                Репетитор по информатике | Мамаев Алексей
            </a>
        </strong>
        <span style="font-size: 16px; color: #333;">
            Разбор <a href='https://euler.jakumo.org/problem/6-raznost-mezdu-summoi-kvadratov-i-kvadratom-summy' style="text-decoration: none; color: #1a73e8;">
            6-й задачи</a> проекта Эйлера
        </span>
        <span style="font-size: 16px; color: #333;">
            <br><br><br><br><br><br>ТГК: <a href='https://t.me/M4m4ev_Alex' style="text-decoration: none; color: #1a73e8;">@M4m4ev_Alex</a><br>
        </span>
    </div>
</div>

### Задача 6

Сумма квадратов первых десяти натуральных чисел равна

```text
1² + 2² + ... + 10² = 385
```

Квадрат суммы первых десяти натуральных чисел равен

```text
(1 + 2 + ... + 10)² = 55² = 3025
```

Следовательно, разность между квадратом суммы и суммой квадратов составляет

```text
3025 − 385 = 2640
```

Необходимо найти разность между **квадратом суммы** и **суммой квадратов** первых **100** натуральных чисел, то есть вычислить значение выражения

```text
(1 + 2 + ... + 100)² − (1² + 2² + ... + 100²)
```

Ответ: 25164150

## 1) Последовательное вычисление суммы

В данном решении обе необходимые величины вычисляются одновременно за один проход по диапазону чисел.

Во время обхода последовательно накапливаются:

- сумма квадратов натуральных чисел;
- сумма натуральных чисел.

После завершения цикла квадрат суммы вычисляется отдельно, а затем находится искомая разность.

### Алгоритм

1. Создаются две переменные:
   - `sk` — сумма квадратов;
   - `ks` — сумма чисел.
2. Последовательно перебираются числа от `1` до `n`.
3. Для каждого числа:
   - квадрат числа добавляется к `sk`;
   - само число добавляется к `ks`.
4. После завершения цикла вычисляется `ks²`.
5. Возвращается разность между квадратом суммы и суммой квадратов.

### Особенности

- Оба значения вычисляются за один проход по диапазону.
- Используется всего две дополнительные переменные.
- Временная сложность составляет `O(n)`.
- Пространственная сложность составляет `O(1)`.

### Вывод

Последовательное вычисление суммы квадратов и суммы чисел позволяет получить решение с линейной сложностью.

In [1]:
from time import perf_counter # измерение времени выполнения алгоритма

In [6]:
def get_result_1(n):
    sk, ks = 0, 0
    for i in range(1, n + 1):
        sk += i ** 2
        ks += i
    return ks ** 2 - sk


start = perf_counter()
print(get_result_1(100))
end = perf_counter()

print(f'{end - start:.5f}')

25164150
0.00446


## 2) Использование математических формул

В данном решении сумма первых `n` натуральных чисел и сумма их квадратов вычисляются напрямую по известным математическим формулам.

Это позволяет полностью отказаться от цикла и получить результат за постоянное количество арифметических операций.

### Квадрат суммы натуральных чисел

Сумма первых `n` натуральных чисел вычисляется по формуле

$$
1 + 2 + 3 + \ldots + n =
\frac{n(n+1)}{2}
$$

По условию задачи требуется квадрат этой суммы:

$$
\left(\sum_{i=1}^{n} i\right)^2
=
\left(\frac{n(n+1)}{2}\right)^2
$$

В программе:

```python
ks = ((1 + n) * n // 2) ** 2
```

---

### Сумма квадратов натуральных чисел

Сумма квадратов первых `n` натуральных чисел вычисляется по формуле

$$
1^2 + 2^2 + 3^2 + \ldots + n^2 =
\frac{n(n+1)(2n+1)}{6}
$$

или

$$
\sum_{i=1}^{n} i^2 =
\frac{n(n+1)(2n+1)}{6}
$$

В программе:

```python
sk = (n * (n + 1) * (2 * n + 1)) // 6
```

---

### Алгоритм

1. По формуле вычисляется сумма первых `n` натуральных чисел.
2. Полученная сумма возводится в квадрат.
3. По второй формуле вычисляется сумма квадратов первых `n` натуральных чисел.
4. Возвращается разность между квадратом суммы и суммой квадратов.

### Особенности

- Циклы не используются.
- Все вычисления выполняются по готовым математическим формулам.
- Временная сложность составляет `O(1)`.
- Пространственная сложность составляет `O(1)`.
- Решение значительно эффективнее последовательного перебора при больших значениях `n`.

### Вывод

Использование математических формул позволяет получить наиболее эффективное решение задачи. Количество выполняемых операций не зависит от значения `n`, поэтому алгоритм имеет постоянную временную сложность.

In [7]:
def get_result_2(n):
    ks = ((1 + n) * n // 2) ** 2
    sk = (n * (n + 1) * (2 * n + 1)) // 6
    return ks - sk


start = perf_counter()
print(get_result_2(100))
end = perf_counter()

print(f'{end - start:.5f}')

25164150
0.00010


## Сравнение времени выполнения

Для сравнения производительности используется функция `timeit()`, предназначенная специально для измерения времени выполнения небольших участков кода.

Каждый алгоритм запускается **1000 раз**, после чего общее время делится на количество запусков. Таким образом получается **среднее время выполнения одного запуска**.

In [12]:
from timeit import timeit

funcs = (
    get_result_1,
    get_result_2,
)

for i, func in enumerate(funcs, start=1):
    tm = timeit(lambda: func(100), number=1000) / 1000
    print(f'Алгоритм {i}: {tm:.7f}')

Алгоритм 1: 0.0000074
Алгоритм 2: 0.0000003


## Заключение

В ходе решения задачи были рассмотрены два различных подхода к вычислению разности между квадратом суммы и суммой квадратов первых `n` натуральных чисел.

Первое решение основано на последовательном вычислении обеих величин за один проход по диапазону чисел. Такой подход отличается простотой реализации и хорошо подходит для понимания идеи задачи, однако время его работы линейно зависит от значения `n`.

Во втором решении используются известные математические формулы для суммы первых натуральных чисел и суммы их квадратов. Это позволяет полностью отказаться от перебора и получить ответ за постоянное количество арифметических операций.

Среднее время выполнения алгоритмов:

- **Алгоритм 1:** `0.0000074` с
- **Алгоритм 2:** `0.0000003` с

Полученные результаты показывают, что использование математических формул значительно повышает производительность алгоритма. Второе решение оказалось примерно в **25 раз быстрее** первого, а его время выполнения практически не зависит от значения `n`.